### PR TITLE
Implement inter-workload dependencies handling in agent

### PR DESCRIPTION
**Pull request by inf17101**
Issues: #48 

The requirement linkage and the refactoring of the runtime manager tests will be done in a separate subsequent PR.

# Definition of Done

The PR shall be merged only if all items mentioned in [CONTRIBUTING.md](https://github.com/eclipse-ankaios/ankaios/blob/main/CONTRIBUTING.md#how-to-contribute) have been followed. In case an item is not applicable as described, please provide a short explanation in the description.

- [x] All steps in [CONTRIBUTING.md](https://github.com/eclipse-ankaios/ankaios/blob/main/CONTRIBUTING.md#how-to-contribute) have been handled

### CODE REVIEW

**Comment by inf17101** [comment]
@krucod3, @christoph-hamm: I think even if it is not 100% ready, you can start reviewing the implementation and the tests. I will add one more stest for an important corner case today and I will change the requirements and adapt the linkage. Maybe some parts I will refactor but you can start and then it is anyway covered with your comments.

**Comment by inf17101** [comment]
Shall we change the Polling Rate in the GenericStateChecker ? Currently it is 1 sec, but we thought about changing to 500ms. Shall we do that or keep it?

**Comment by inf17101** [comment]
Maybe I rename the `DependencyState` enum into `WorkloadState` enum, because it describes when a workload is ready or pending not a "dependency".

File: agent/src/runtime_manager.rs [hunk 1]
```diff
@@ -1,22 +1,41 @@
+// Copyright (c) 2023 Elektrobit Automotive GmbH
```
[the remark this answers is not in the document]
Comment (inf17101) [reply agent/src/runtime_manager.rs:1]: I would not change this to 2024 because the file was created in 2023. It was just missing and I noticed that during development.

File: tests/stests/workloads/interworkload_dependencies.robot [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+Test Ankaios observes the inter-workload dependencies when creating workloads
+    [Documentation]    Perform a create of an workload only if its start dependencies are fulfilled.
+    [Setup]    Run Keywords    Setup Ankaios
+    # Preconditions
+    # This test assumes that all containers in the podman have been created with this test -> clean it up first
+    Given Podman has deleted all existing containers
+    And Ankaios server is started with config "${CONFIGS_DIR}/create_workloads_with_dependencies_config.yaml"
+    And Ankaios agent is started with name "agent_A"
+    And Ankaios agent is started with name "agent_B"
+    # Actions
```
[the remark this answers is not in the document]
Comment (inf17101) [reply tests/stests/workloads/interworkload_dependencies.robot:39]: I think the whole Ankaios startup is the action here the user triggers. So I would move also the agents down into the action section. The "podman has deleted all existing containers" is then the single precondition.

File: agent/src/workload_state_sender.rs [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+        execution_state: ExecutionState,
+    ) -> Result<(), String>;
+
+    async fn report_starting(&self, instance_name: &WorkloadInstanceName);
+    async fn report_stopping(&self, instance_name: &WorkloadInstanceName);
+    async fn report_stopping_failed(&self, instance_name: &WorkloadInstanceName);
+    async fn report_removed(&self, instance_name: &WorkloadInstanceName);
+}
+
+#[async_trait]
```
[the remark this answers is not in the document]
Comment (inf17101) [reply agent/src/workload_state_sender.rs]: Fixed

File: common/src/objects/mod.rs [hunk 4]
```diff
@@ -19,6 +19,21 @@
...
+#[cfg(any(feature = "test_utils", test))]
+pub use stored_workload_spec::{
+    generate_test_stored_workload_spec, generate_test_stored_workload_spec_with_config,
+};
+
+pub use stored_workload_spec::StoredWorkloadSpec;
+
+// mod external_state;
+
+// pub use external_state::ExternalCompleteState;
```
[the remark this answers is not in the document]
Comment (inf17101) [reply common/src/objects/mod.rs]: Removed

File: common/src/objects/mod.rs [hunk 5]
```diff
@@ -19,6 +19,21 @@
...
+
+mod stored_workload_spec;
+#[cfg(any(feature = "test_utils", test))]
+pub use stored_workload_spec::{
+    generate_test_stored_workload_spec, generate_test_stored_workload_spec_with_config,
+};
+
+pub use stored_workload_spec::StoredWorkloadSpec;
+
+// mod external_state;
```
[the remark this answers is not in the document]
Comment (inf17101) [reply common/src/objects/mod.rs]: Removed

File: server/src/main.rs [hunk 6]
```diff
@@ -48,8 +47,9 @@ async fn main() {
                 fs::read_to_string(config_path).unwrap_or_exit("Could not read the startup config");
             // [impl->swdd~server-state-in-memory~1]
             // [impl->swdd~server-loads-startup-state-file~2]
-            let state: State =
-                state_parser::parse(data).unwrap_or_exit("Parsing start config failed with error");
+            let state: State = serde_yaml::from_str(&data)
+                // state_parser::parse(data)
```
[the remark this answers is not in the document]
Comment (inf17101) [reply server/src/main.rs]: Removed

File: agent/src/workload_scheduler/scheduler.rs [hunk 7]
```diff
@@ -0,0 +1,1482 @@
...
+#[cfg_attr(test, automock)]
+impl WorkloadScheduler {
+    pub fn new(workload_state_tx: ToServerSender) -> Self {
+        WorkloadScheduler {
+            queue: WorkloadOperationQueue::new(),
+            workload_state_sender: workload_state_tx,
+        }
+    }
+
+    async fn report_pending_create_state(&self, pending_workload: &WorkloadSpec) {
```
Comment (christoph-hamm) [inline agent/src/workload_scheduler/scheduler.rs:56]: Please change the vertical ordering. The public methods should be at the top and functions calls should always go downward.

**Review by christoph-hamm — CHANGES_REQUESTED** [review]

File: agent/src/workload_scheduler/scheduler.rs (hunk 7 above)
Comment (inf17101) [reply agent/src/workload_scheduler/scheduler.rs:56]: Thx, Done.

File: agent/src/workload_operation.rs [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+
+#[derive(Debug, Clone, PartialEq)]
+pub enum WorkloadOperation {
+    Create(WorkloadSpec),
+    Update(WorkloadSpec, DeletedWorkload),
+    UpdateDeleteOnly(DeletedWorkload),
+    Delete(DeletedWorkload),
+}
+
+pub type WorkloadOperations = Vec<WorkloadOperation>;
```
[the remark this answers is not in the document]
Comment (inf17101) [reply agent/src/workload_operation.rs:25]: I think this type was planned to be enhanced or to have a value at later. For now it is not needed anymore. Thank you. I will remove that.

[the remark this answers is not in the document]
Comment (inf17101) [reply agent/src/workload_operation.rs]: Done

File: common/src/objects/state.rs [hunk 9]
```diff
@@ -1,4 +1,4 @@
-// Copyright (c) 2023 Elektrobit Automotive GmbH
+// Copyright (c) 2024 Elektrobit Automotive GmbH
```
[the remark this answers is not in the document]
Comment (inf17101) [reply common/src/objects/state.rs:1]: According to the commits I have not changed it, but I will revert it quickly.

File: common/src/objects/workload_state.rs [hunk 10]
```diff
@@ -291,6 +344,20 @@ impl ExecutionState {
...
+        ExecutionState {
+            state: ExecutionStateEnum::Pending(PendingSubstate::WaitingToStart),
+            additional_info: "waiting for workload dependencies.".to_string(),
+        }
+    }
+
+    pub fn waiting_to_stop() -> Self {
+        ExecutionState {
+            state: ExecutionStateEnum::Stopping(StoppingSubstate::WaitingToStop),
+            additional_info: "waiting for workload dependencies.".to_string(),
```
[the remark this answers is not in the document]
Comment (inf17101) [reply common/src/objects/workload_state.rs:358]: I think it would be better if the string shows the names of the workloads it waits for (for waiting to stop and for waiting to start).
And the Ank CLI can limit the amount of characters if a workload has too much dependency names.

What do you think?

File: common/src/objects/state.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (inf17101) [reply common/src/objects/state.rs]: Done

File: agent/src/runtime_manager.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (lingnoi) [reply agent/src/runtime_manager.rs:1]: Ok, for me, but I just ask myself why do we need the copyright statement then, when the year has been taken from the creation date of the file.

File: agent/src/workload_scheduler/dependency_state_validator.rs [hunk 11]
```diff
@@ -0,0 +1,239 @@
...
+    #[test]
+    fn utest_create_fulfilled() {
+        let workload_with_dependencies = generate_test_workload_spec_with_dependencies(
+            AGENT_A,
+            WORKLOAD_NAME_1,
+            RUNTIME,
+            HashMap::from([(WORKLOAD_NAME_2.to_string(), AddCondition::AddCondRunning)]),
+        );
+
+        let execution_state = Box::leak(Box::new(Some(ExecutionState::running())));
```
Comment (lingnoi) [inline agent/src/workload_scheduler/dependency_state_validator.rs:90]: As it is used in the utests only and the size is small leaking memory for comfort is ok in this case.

**Review by lingnoi — APPROVED** [review]
LGTM, now 👍

File: common/src/objects/workload_state.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (christoph-hamm) [reply common/src/objects/workload_state.rs:358]: This would be nice. But if this addition info is not updated for already running dependencies, this could be misleading and updating this additional info would be to much work. Just remove the additional_info.

**Comment by inf17101** [comment]
The introduced manual mock in [2fe1430](https://github.com/eclipse-ankaios/ankaios/pull/189/commits/2fe1430fad3b870970934e853e454cd810f46177) looks good to me.
LGTM 👍

**Review by krucod3 — APPROVED** [review]
LGTM
Added a comment about one misleading trace, but to avoid conflicts, I'll pack the changes to my PR.

File: agent/src/agent_manager.rs [hunk 12]
```diff
@@ -26,85 +28,150 @@ pub struct AgentManager {
...
+            std::unreachable!("expected UpdateWorkloadState msg.");
+        };
+
+        workload_states.iter().for_each(|new_workload_state| {
+            log::info!(
+                "The agent '{}' reports workload state '{:?}' for the workload '{}'",
+                new_workload_state.instance_name.agent_name(),
+                new_workload_state.execution_state,
+                new_workload_state.instance_name.workload_name(),
+            );
```
[the remark this answers is not in the document]
Comment (krucod3) [reply agent/src/agent_manager.rs:158]: I'll do the changes in #191
